File: benchmarks/lamutex/Python/lamutex.py
#!/usr/bin/env python3
import sys
import time
import json
import queue
import random
import resource
import threading
import multiprocessing
from heapq import heappush, heappop, heapify

BROADCAST = -1
LOG = -2

# ==============================
# The 'Site' class represents a peer in the network. Each 'Site' is run in a
# separate process
# ==============================
class Site(multiprocessing.Process):
    # ====================
    # Each 'Site' will have one 'Comm' thread, its sole purpose is to reply all
    # incoming requests:
    # ====================
    class Comm(threading.Thread):
        def __init__(self, pid, net, queue):
            threading.Thread.__init__(self)
            self.peerid = pid   # self's unique process id
            self.queue = queue  # Queue used to communicate with our main Site
                                # process
            self.net = net      # Interface to the "network" (which is a
                                # dictionary containing an 'in' queue and an
                                # 'out' queue.)
            self.daemon = True

        # Simple algorithm for Comm: Take a message from net 'in' and put it
        # into the queue for Site. If the message is a request we reply it:
        def run(self):
            while True:
                mtype, source, clock = receive(self.net)
                self.queue.put((mtype, source, clock))
                if (mtype == "request"):
                    send(self.net, source, ("ack", self.peerid, clock))

    def __init__(self, pid, net, iterations, peers):
        multiprocessing.Process.__init__(self)
        self.peerid = pid       # Our unique peer id
        self.net = net          # Interface to the "network" (which is a dict
                                # containing an in queue and an out queue.)
        self.peers = peers      # List of all processes, including self
        self.iterations = iterations      # Number of iterations to run
        self.thread_queue = queue.Queue() # Queue used to communicate with Comm
        self.acks = set()                 # The set of replies we've received
        self.reqs = []                    # Request priority queue, in clock order
        self.clock = 0                    # Our logical clock
        self.comm = Site.Comm(pid, net, self.thread_queue)
        self.running = False

    def run(self):
        self.comm.start()       # Start the "Comm" thread
        while not self.running:
            self.handle_message(True)

        rudata_start = resource.getrusage(resource.RUSAGE_SELF)
        count = 0
        while(self.running and count < self.iterations):
            # Non CS:
            self.label()        # labels mark the points where we can "break"
            self.work()
            self.label()
            self.enter_critical_section()
            print("Process %d(clock:%d) has entered critical section."%
                  (self.peerid,self.clock))
            self.label()
            self.work()
            self.label()
            print("Process %d is leaving critical section."%self.peerid)
            self.leave_critical_section()
            count += 1

        rudata_end = resource.getrusage(resource.RUSAGE_SELF)
        utime = rudata_end.ru_utime - rudata_start.ru_utime
        stime = rudata_end.ru_stime - rudata_start.ru_stime
        send(self.net, LOG, (utime, stime, rudata_end.ru_maxrss))
        while self.running:
            self.handle_message(True)

    # This simulates the controlled "label" mechanism. Currently we simply
    # handle one message on one label call:
    def label(self):
        self.handle_message(False)

    # Handles one message. 'block' indicates whether to blocking waiting for
    # next message to come in if the queue is currently empty:
    def handle_message(self, block):
        try:
            (mtype, source, clock) = self.thread_queue.get(block)
            if mtype == "request":
                # Put request on queue and update clock
                heappush(self.reqs, (clock, source))
                self.clock = max(self.clock, clock) + 1
            elif mtype == "ack":
                self.acks.add(source)
            elif mtype == "release":
                # Remove the sender from our request queue
                for (clk, peer) in self.reqs:
                    if peer == source:
                        self.reqs.remove((clk, peer))
                        # Note that we have to re-heapify after the remove:
                        heapify(self.reqs)
            elif mtype == "terminate":
                self.running = False
            elif mtype == "start":
                self.running = True
            else:
                raise RuntimeError("Unknown message type " + mtype)
        except queue.Empty:
            pass

    # We keep handling messages until the wait condition becomes true:
    def enter_critical_section(self):
        self.acks = set()
        send(self.net, BROADCAST, ("request", self.peerid, self.clock))
        while (len(self.reqs) == 0 or
               self.reqs[0][1] != self.peerid or
               len(self.acks) != len(self.peers)):
            self.handle_message(True)

    def leave_critical_section(self):
        send(self.net, BROADCAST, ("release", self.peerid, self.clock))

    # Simulate work, waste some random amount of time:
    def work(self):
#        time.sleep(random.randint(1, 5))
        return


# ==============================
# The "Simulator" class takes care of message routing between the peers:
# ==============================
class Simulator(threading.Thread):
    def __init__(self, sites, netq):
        threading.Thread.__init__(self)
        self.sites = sites
        self.netq = netq    # This is the shared 'out' queue of all the
                            # processes

        self.perf_usrtime = 0
        self.perf_systime = 0
        self.perf_memory = 0

        self.completed = 0

    def run(self):
        t1 = time.perf_counter()
        for s in self.sites:
            s.net['in'].put(('start', -1, -1))

        while self.completed < len(self.sites):
            message = self.netq.get(True)
            self.dispatch_message(message)

        t2 = time.perf_counter()
        for s in self.sites:
            s.net['in'].put(('terminate', -1, -1))

        jsondata = {'Wallclock_time': t2 - t1,
                    'Total_processes': self.completed,
                    'Total_process_time': self.perf_usrtime + self.perf_systime,
                    'Total_user_time': self.perf_usrtime,
                    'Total_memory': self.perf_memory}
        jsonoutput = json.dumps(jsondata)
        print("###OUTPUT: " + jsonoutput)

        for site in self.sites:
            site.join()

    def dispatch_message(self, message):
        (to, m) = message
        if to == BROADCAST:
             for site in self.sites:
                 site.net['in'].put(m)
        elif to == LOG:
            self.completed += 1

            utime, stime, maxrss = m
            self.perf_usrtime += utime
            self.perf_systime += stime
            self.perf_memory += maxrss

        else:
            self.sites[to].net['in'].put(m)

# Create nproc "Sites":
def create_sites(nproc, q, iterations):
    return [Site(id, {'out':q, 'in':multiprocessing.Queue()},
                 iterations, range(0, nproc))
            for id in range(0, nproc)]

# ==============================
# Main program entry point:
# ==============================
def main(nproc, iterations):
    queue = multiprocessing.Queue()
    print("Creating sites...")
    sites = create_sites(nproc, queue, iterations)
    print("Creating simulator...")
    sim = Simulator(sites, queue)
    print("Starting %d sites...\n"%nproc)
    for site in sites:
        site.start()
    print("Starting simulator...")
    sim.start()
    sim.join()

# Wrapper functions for message passing:
def send(net, to, message):
    net['out'].put((to, message))

def receive(net):
    return net['in'].get()

if __name__ == "__main__":
    nprocs = int(sys.argv[1]) if len(sys.argv) > 1 else 10
    iterations = int(sys.argv[2]) if len(sys.argv) > 2 else 5
    main(nprocs, iterations)

File: benchmarks/lamutex/Erlang/lamutex.erl
%% Copyright (c) 2010-2015 Bo Lin
%% Copyright (c) 2010-2015 Yanhong Annie Liu
%% Copyright (c) 2010-2015 Stony Brook University
%% Copyright (c) 2010-2015 The Research Foundation of SUNY

%% Permission is hereby granted, free of charge, to any person
%% obtaining a copy of this software and associated documentation files
%% (the "Software"), to deal in the Software without restriction,
%% including without limitation the rights to use, copy, modify, merge,
%% publish, distribute, sublicense, and/or sell copies of the Software,
%% and to permit persons to whom the Software is furnished to do so,
%% subject to the following conditions:

%% The above copyright notice and this permission notice shall be
%% included in all copies or substantial portions of the Software.

%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
%% EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
%% MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
%% NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
%% LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
%% OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
%% WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

-module(lamutex).
-export([run/2, start/1, start/0, run_processes/2]).
-export([site1/2, site/2, comm/5, broadcast_tracker/2]).

%% ------------------------------
%% Command line entry point
%% ------------------------------
start() ->
    run(10, 10).

start([NPeers]) ->
    run(list_to_integer(NPeers), 5);
start([NPeers, RTime]) ->
    run(list_to_integer(NPeers), list_to_integer(RTime)).

%% ------------------------------
%% Program entry point
%% ------------------------------
run(NumPeers, RunTime)
  when is_integer(NumPeers), is_integer(RunTime) ->
    io:format("Starting Comm processes...\n"),
    Comms = create_comm_processes(NumPeers),
    %% Register the broadcast process so Comms would know where to send
    %% broadcast requests:
    register(tracker, spawn(lamutex, broadcast_tracker, [Comms, self()])),
    io:format("Starting Site processes... ~w~n \n", [length(Comms)]),
    Sites = create_site_processes(Comms, RunTime),
    statistics(runtime),
    {WallclockTime, _Result} = timer:tc(lamutex, run_processes, [Sites, NumPeers]),
    {_, CpuTime} = statistics(runtime),
    dump_perf_data(WallclockTime, NumPeers, CpuTime),
    halt(0).

run_processes(Sites, NumPeers) ->
    do_broadcast(Sites, start),
    wait_all_done(NumPeers).

wait_all_done(0) ->
    ok;
wait_all_done(NumLeft)
  when NumLeft > 0 ->
    receive
        {finished, _Id} ->
            wait_all_done(NumLeft-1)
    end.

get_memory_stats() ->
    {ok, StatsFile} = file:open(filename:join(['/proc', os:getpid(), 'status']),
                                [read]),
    {ok, Matcher} = re:compile("^VmHWM:[^0-9]*\([0-9]+\) kB"),
    case first_matching_line(Matcher, StatsFile) of
        {vmhwm, Value} -> Value;
        notfound -> "0"
    end.

first_matching_line(Matcher, Device) ->
    case io:get_line(Device, "") of
        eof ->
            file:close(Device),
            notfound;
        Line ->
            case re:run(Line, Matcher, [{capture, all_but_first, list}]) of
                {match, [Cap]} ->
                    file:close(Device),
                    {vmhwm, Cap};
                nomatch ->
                    first_matching_line(Matcher, Device)
            end
    end.

dump_perf_data(WallclockTime, NumPeers, CpuTime) ->
    io:format("###OUTPUT: {\"Total_memory\": ~s, "
              "\"Wallclock_time\": ~f, \"Total_processes\": ~w, "
              "\"Total_process_time\": ~f, \"Total_user_time\": ~f}~n",
              [get_memory_stats(), WallclockTime/1000000,
               NumPeers, CpuTime/1000, CpuTime/1000]).

%% --------------------
%% This helper process handles all broadcast requests. It's a small hackery to
%% get around the circular dependency of letting every peer process know about
%% all other peers before they are actually created. Maybe there's a better
%% way to do this...
%% --------------------
broadcast_tracker(Targets, ParentId) ->
    receive
        {broadcast, Message} ->
            do_broadcast(Targets, Message),
            broadcast_tracker(Targets, ParentId);
        {finished, Time} ->
            ParentId ! {finished, Time},
            broadcast_tracker(Targets, ParentId)
    end.
do_broadcast([], _) ->
    ok;
do_broadcast([To|Rest], Message) ->
    To ! Message,
    do_broadcast(Rest, Message).


%% --------------------
%% Create all the "Comm" processes.
%% --------------------
create_comm_processes(N) ->
    [spawn(lamutex, comm, [Id, N, [], 0, ignore]) || Id <- lists:seq(1, N)].

%% --------------------
%% Create all the "Site" processes.
%% --------------------
create_site_processes(CommList, Rounds) ->
    [spawn(lamutex, site1, [Id, Rounds]) || Id <- CommList].

%% --------------------
%% Site entry point.
%% --------------------
site1(CommId, Rounds) ->
    io:format("Site ~w starting...\n", [CommId]),
    receive
        start -> ok
    end,
    site(CommId, Rounds),
    io:format("Process ~w done. \n", [CommId]),
    tracker ! {finished, CommId}.

%% --------------------
%% The "Site" process is now extremely simple and straight forward: 1. do
%% non-critical stuff; 2. try enter critical section by notifying our "Comm"
%% and waiting for the "goahead" signal; 3. do critical work; 4. leave CS. It
%% maintains no data apart from the corresponding "Comm" PID that handles CS
%% requests on its behalf.
%% --------------------
site(_CommId, 0) ->
    ok;
site(CommId, Rounds) ->
    %% Non CS:
    enter_critical_section(CommId),
    %% In CS:
    io:format("In critical section: ~w \n", [CommId]),
    %procrastinate(),
    %% Leave CS:
    io:format("Leaving critical section: ~w \n", [CommId]),
    leave_critical_section(CommId),
    site(CommId, Rounds-1).

%% --------------------
%% All data structures pertaining to critical section management is now moved
%% into the "Comm" process.
%% --------------------
comm(SiteId, NPeers, ReqQueue, Clock, AckSet) ->
    %% First check whether our Site is ready to enter CS:
    Result = check_can_enter_cs(self(), ReqQueue, AckSet, NPeers),
    if
        Result ->
            SiteId ! goahead,
            comm(SiteId, NPeers, ReqQueue, Clock, ignore);
        true ->
            ok
    end,
    %% Then handle the next message:
    receive 
        {request, From, Clk} ->
            From ! {ack, self(), Clock},        % Send ack reply
            comm(SiteId, NPeers,
                 lists:keystore(From, 2, ReqQueue, {Clk, From}), 
                 max(Clock, Clk) + 1,
                 AckSet);

        {ack, From, _Clk} ->
            %% 'ignore' is used as a minor optimiazation: If our Site is not
            %% trying to enter CS, then we can safely ignore all ack messages:
            if
                AckSet == ignore ->
                    comm(SiteId, NPeers, ReqQueue, Clock, AckSet);
                true ->
                    comm(SiteId, NPeers, ReqQueue, Clock,
                         sets:add_element(From, AckSet))
            end;

        {release, From, _Clk} ->
            comm(SiteId, NPeers,
                 lists:keydelete(From, 2, ReqQueue), Clock, AckSet);

        %% These messages are used to communicate with our Site:
        {enter, Id} -> % Our own site wishes to enter
            broadcast({request, self(), Clock}),
            comm(Id, NPeers, ReqQueue, Clock, sets:new());
        {release, Id} when Id == SiteId-> % Our own site is done
            broadcast({release, self(), Clock}),
            comm(Id, NPeers, ReqQueue, Clock, ignore)
    end.

%% ==================================================
%% Helpers
%% ==================================================

%% Just send the message to the broadcast tracker process:
broadcast(Message) ->
    tracker ! {broadcast, Message}.

%% CS condition check. If successful send goahead back to site:
check_can_enter_cs(CommId, ReqQueue, AckSet, NumPeers) ->
    AckSet /= ignore andalso
        sets:size(AckSet) == NumPeers andalso
        length(ReqQueue) > 0 andalso
        element(2, lists:min(ReqQueue)) == CommId.

%% Time waster:
procrastinate() ->
    receive
    after random:uniform(3) * 1000 ->
            ok
    end.

enter_critical_section(CommId) ->
    CommId ! {enter, self()},
    receive
        goahead ->
            ok
    end.

leave_critical_section(CommId) ->
    CommId ! {release, self()}.
